alf/algorithms/vae.py

```diff
@@ -179,3 +184,143 @@ def train_step(self, inputs, state=()):
+        Neurons for Conditional Computation", 2013.
+
+    In short, we can re-parameterize the one-hot latent embedding :math:`z` as
+
+    .. math::
+
+        \hat{z} = z + z_{prob} - SG(z_{prob})
+
+    Because :math:`z` is a sampled discrete variable, it has no gradient. So
+    the parameter gradient is
+
+    .. math::
+
+        \frac{\partial L}{\partial \hat{z}}\frac{\partial \hat{z}}{\partial \theta}
+        = \frac{\partial L}{\partial \hat{z}}\frac{\partial z_{prob}}{\partial \theta}
+
+    For implementation, we directly use ``OnehotCategoricalProjectionNetwork`` to
+    model the posterior distribution.
+    """
+
+    def __init__(self,
+                 z_spec: BoundedTensorSpec,
+                 input_tensor_spec: alf.NestedTensorSpec = None,
+                 encoder_cls: Callable = EncodingNetwork,
+                 prior_input_tensor_spec: alf.NestedTensorSpec = None,
+                 prior_encoder_cls: Callable = None,
+                 beta: float = 1.,
+                 target_kld_per_categorical: float = None,
+                 beta_optimizer: torch.optim.Optimizer = None,
+                 name: str = "DiscreteVAE"):
+        """
+        Args:
+            z_spec: a rank-0 or rank-1 tensor spec for the discrete posterior.
+                If it's rank-1, then ``z`` represents multiple  Categorials.
+            input_tensor_spec: the input spec.
+            encoder_cls: an encoding network to
+                preprocess input data before projecting it into a discrete
+                distribution. If ``prior_encoder_cls`` is None, this network must
+                handle input with spec ``input_tensor_spec``. If ``prior_encoder_cls``
+                is not None, this network must be handle input with spec
+                ``(prior_input_tensor_spec, input_tensor_spec, prior_encoder.output_spec)``.
+            prior_input_tensor_spec: the input spec for the prior encoder.
+            prior_encoder_cls: an encoding network that outputs an embedding to
+                be projected into a prior ``z`` distribution given the prior input.
+            beta: the weight for KL-divergence
+            target_kld_per_categorical: if not None, then this will be used as the
+                target KLD per Categorical to automatically tune beta.
+            beta_optimizer: if not None, will be used to train beta.
+            name (str):
+        """
+        Algorithm.__init__(self, name=name)
+
+        assert z_spec.is_discrete
+
+        prior_z_network = None
+        if prior_encoder_cls is not None:
+            prior_encoder = prior_encoder_cls(
+                input_tensor_spec=prior_input_tensor_spec)
+            prior_proj_net = OnehotCategoricalProjectionNetwork(
+                input_size=prior_encoder.output_spec.numel, action_spec=z_spec)
+            prior_z_network = alf.nn.Sequential(prior_encoder, prior_proj_net)
+            logits_spec = TensorSpec((np.prod(prior_proj_net._output_shape), ))
```

Comment:
do you think the code will be cleaner if we simply require that encoder and prior_encoder output the logits directly? And it will be quite easy to support Gumbel softmax.

Reply:
Gumbel softmax seems not what I need. For discrete VAEs we would expect the sampled latent code to be completely discrete (hard labels). Gumbel softmax will still produce soft labels due to using softmax (with a temperature) to replace argmax. 

![image](https://user-images.githubusercontent.com/51248379/186258315-d9db7a64-aa96-4608-9dbf-3be9e13a704f.png)

Reply:
As in the paper, It can also combined with the straight through trick

Reply:
> As in the paper, it can also be combined with the straight through trick

I'm actually a bit confused about combining Gumbel softmax and straight-through. The Gumbel trick is basically equivalent to Categorical sampling. If eventually we'll use straight-through on the one-hot, why not bother using Gumbel softmax to replace the categorical sampling in the first place? 

![image](https://user-images.githubusercontent.com/51248379/186259723-683ba1b9-4bd7-4a8f-b2df-a4b0bf8fef68.png)

Reply:
We could in fact support Gumbel softmax for the latent code, but comment on it to be softly sparse (depending on the temperature) instead of hard onehot. For hard latent code, still use the current onehot straight-through, as Gumbel softmax+straight-through seems unnecessary to me.

Reply:
never mind, I think I understand now that there is still some difference between the two and the paper claims ST gumbel softmax to have lower variance than plain ST.